tests/api/report-controllers.php

```diff
@@ -0,0 +1,107 @@
+	/**
+	 * Setup test admin notes data. Called before every test.
+	 *
+	 * @since 3.5.0
+	 */
+	public function setUp() {
+		parent::setUp();
+
+		$this->user = $this->factory->user->create(
+			array(
+				'role' => 'administrator',
+			)
+		);
+
+		wp_set_current_user( $this->user );
+	}
+
+	/**
+	 * Ensure that every REST controller works with their defaults.
+	 */
+	public function test_default_params() {
+		// These controllers intentionally missing:
+		// - `WC_Admin_REST_Setting_Options_Controller`
+		// - `WC_Admin_REST_Data_Download_Ips_Controller`
+		// because they don't have defaults for required params.
+		$controller_classes = array(
+			'WC_Admin_REST_Admin_Notes_Controller',
+			'WC_Admin_REST_Admin_Note_Action_Controller',
+			'WC_Admin_REST_Coupons_Controller',
+			'WC_Admin_REST_Customers_Controller',
+			'WC_Admin_REST_Data_Controller',
+			'WC_Admin_REST_Data_Countries_Controller',
+			'WC_Admin_REST_Leaderboards_Controller',
+			'WC_Admin_REST_Orders_Controller',
+			'WC_Admin_REST_Products_Controller',
+			'WC_Admin_REST_Product_Categories_Controller',
+			'WC_Admin_REST_Product_Variations_Controller',
+			'WC_Admin_REST_Product_Reviews_Controller',
+			'WC_Admin_REST_Product_Variations_Controller',
+			'WC_Admin_REST_Reports_Controller',
+			'WC_Admin_REST_Reports_Import_Controller',
+			'WC_Admin_REST_Reports_Products_Controller',
+			'WC_Admin_REST_Reports_Variations_Controller',
+			'WC_Admin_REST_Reports_Products_Stats_Controller',
+			'WC_Admin_REST_Reports_Revenue_Stats_Controller',
+			'WC_Admin_REST_Reports_Orders_Controller',
+			'WC_Admin_REST_Reports_Orders_Stats_Controller',
+			'WC_Admin_REST_Reports_Categories_Controller',
+			'WC_Admin_REST_Reports_Taxes_Controller',
+			'WC_Admin_REST_Reports_Taxes_Stats_Controller',
+			'WC_Admin_REST_Reports_Coupons_Controller',
+			'WC_Admin_REST_Reports_Coupons_Stats_Controller',
+			'WC_Admin_REST_Reports_Stock_Controller',
+			'WC_Admin_REST_Reports_Stock_Stats_Controller',
+			'WC_Admin_REST_Reports_Downloads_Controller',
+			'WC_Admin_REST_Reports_Downloads_Stats_Controller',
+			'WC_Admin_REST_Reports_Customers_Controller',
+			'WC_Admin_REST_Reports_Customers_Stats_Controller',
+			'WC_Admin_REST_Taxes_Controller',
+			'WC_Admin_REST_Onboarding_Levels_Controller',
+			'WC_Admin_REST_Onboarding_Profile_Controller',
+			'WC_Admin_REST_Onboarding_Plugins_Controller',
+			'WC_Admin_REST_Reports_Performance_Indicators_Controller',
+		);
+
+		// Force REST controllers to load.
+		do_action( 'rest_api_init' );
+
+		foreach ( $controller_classes as $controller_class ) {
+			$controller = new $controller_class();
+			$params     = $controller->get_collection_params();
+			$defaults   = array();
+
+			foreach ( $params as $arg => $options ) {
+				if ( isset( $options['default'] ) ) {
+					$defaults[ $arg ] = $options['default'];
+				}
+			}
+
+			$request = new WP_REST_Request();
+			$request->set_default_params( $defaults );
```

Comment:
I'm not entirely sure what `set_default_params` does and the [docs](https://developer.wordpress.org/reference/classes/wp_rest_request/set_default_params/) aren't too descriptive, but isn't the purpose of the test to how the endpoint responds without any parameters?

Reply:
It's part of what the `WP_REST_Server` class does when [handling routes](https://github.com/WordPress/WordPress/blob/master/wp-includes/rest-api/class-wp-rest-server.php#L882) - since I wasn't using the server in test, I had to manually apply the defaults.

I've switched the test up to make use of the server instead. Hopefully it's easier to grok: da7754bc